oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java

```diff
@@ -0,0 +1,261 @@
+                boolean shouldSplit = (readPos > splitThreshold);
+                if (shouldSplit && canSplit(splitNodeTypesName, nodeTypeNameStack)) {
+                    writer.close();
+                    LOG.info("created split flat file {} with size {}", currentFile.getAbsolutePath(), FileUtils.byteCountToDisplaySize(currentFile.length()));
+                    readPos = 0;
+                    outFileIndex++;
+                    currentFile = new File(workDir, "split-" + outFileIndex + "-" + getSortedStoreFileName(algorithm));
+                    writer = createWriter(currentFile, algorithm);
+                    splitFlatFiles.add(currentFile);
+                    LOG.info("split position found at line {}, creating new split file {}", lineCount, currentFile.getAbsolutePath());
+                }
+                writer.append(line);
+                writer.newLine();
+                readPos += line.length() + 1;
+                lineCount++;
+            }
+            writer.close();
+            LOG.info("created split flat file {} with size {}", currentFile.getAbsolutePath(), FileUtils.byteCountToDisplaySize(currentFile.length()));
+
+            LOG.info("split total line count: {}", lineCount);
+        }
+
+        if (deleteOriginal) {
+            LOG.info("removing original flat file {} after splitting into {} files", flatFile.getAbsolutePath(), splitFlatFiles);
+            flatFile.delete();
+        }
+
+        return splitFlatFiles;
+    }
+
+    private void updateNodeTypeStack(Stack<String> parentNodeTypeNames, String line) {
+        NodeStateHolder ns = new SimpleNodeStateHolder(line);
+        List<String> pathElements = ns.getPathElements();
+        int currentLineDepth = pathElements.size();
+        int parentTypesDepth = parentNodeTypeNames.size();
+        if (currentLineDepth > parentTypesDepth) {
+            parentNodeTypeNames.add(getJCRPrimaryType(line));
+        } else {
+            int popSize = parentTypesDepth - currentLineDepth + 1;
+            if (parentTypesDepth > 0) {
+                for (int i = 0; i < popSize; i++) {
+                    parentNodeTypeNames.pop();
+                }
+            }
+            parentNodeTypeNames.add(getJCRPrimaryType(line));
+        }
+    }
+
+    private String getJCRPrimaryType(String line) {
+        NodeStateEntry nse = entryReader.read(line);
+        PropertyState property = nse.getNodeState().getProperty(JCR_PRIMARYTYPE);
+        if (property == null) {
+            return "";
+        }
+        Type<?> type = property.getType();
+        if (type == Type.NAME) {
+            return property.getValue(Type.NAME);
+        }
+        return "";
+    }
+
+    private boolean canSplit(Set<String> nodeTypes, Stack<String> nodeTypeNameStack) {
+        if (nodeTypeNameStack.contains("")) {
+            return false;
+        }
+        for (String parentNodeTypeName : nodeTypeNameStack.subList(0, nodeTypeNameStack.size()-1)) {
```

Comment:
this loop is a bit inconsistent to the previous check (```nodeTypeNameStack.contains("")```), because it does not iterate over the last element in the stack, but the ```contains()``` does); is this intendend and valid?

Reply:
The code itself is correct (at least this line is correct), but it is not properly documented.